src/librustc_codegen_llvm/debuginfo/mod.rs

```diff
@@ -279,7 +279,8 @@ pub fn create_function_debug_context(
         }
         None => {}
     };
-    if cx.layout_of(sig.output()).abi == ty::layout::Abi::Uninhabited {
+    // Tell LLVM that functions that return uninhabited types will not return.
+    if sig.output().conservative_is_uninhabited() {
```

Comment:
isn't this way more conservative than the layout check? I think this should stay being the layout check. It does not influence type checking, just llvm optimizations